eclipse-che/che-operator
# feat: configure workspace start timeout from Che Cluster CR

### What does this PR do?
This PR allows the configuration of the DWOC's `progressTimeout` field from the Che Cluster CR's `spec.devEnvironments.startTimeoutSeconds` field.

The `spec.devEnvironments.startTimeoutSeconds` field is a `int32` pointer so that we can differentiate between when the field is unset (`nil`, allowing the default value of 5 minutes to be used) and when a timeout of 0 seconds is desired.

**Note:** This PR also includes the changes from https://github.com/eclipse-che/che-operator/pull/1549 for convenience. I can remove the commits from that PR if desired.

### Screenshot/screencast of this PR
n/a


### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/21820


### How to test this PR?
Tests were added for this PR.

To manually test:

1. Checkout this PR
2. Deploy Che: `build/scripts/olm/test-catalog-from-sources.sh` (for OpenShift)
3. Edit the Che Cluster CR and add the `devEnvironments.startTimeout` field: `kubectl edit checluster eclipse-che -n eclipse-che`

```diff
Spec:
(...)
  Dev Environments:
    Default Components:
      Container:
        Image:           quay.io/devfile/universal-developer-image:ubi8-38da5c2
        Source Mapping:  /projects
      Name:              universal-developer-image
    Default Editor:      che-incubator/che-code/insiders
    Default Namespace:
      Auto Provision:                      true
      Template:                            <username>-che
    Disable Container Build Capabilities:  true
    Seconds Of Inactivity Before Idling:   1800
    Seconds Of Run Before Idling:          -1
+    startTimeoutSeconds:                         420 #7 minutes
    Storage:
      Pvc Strategy:  per-user
  Git Services:
  Networking:
    Auth:
      Gateway:
        Config Labels:
          App:        che
          Component:  che-gateway-config
```
4. Verify that `config.workspace.progressTimeout` in the DWOC has the same value as the `devEnvironments.startTimeout` field from the Che Cluster CR: `kubectl describe dwoc -n eclipse-che`

```diff
Name:         devworkspace-config
Namespace:    eclipse-che
Labels:       <none>
Annotations:  <none>
API Version:  controller.devfile.io/v1alpha1
Config:
  Workspace:
+    Progress Timeout:  420s
(...)
```
5. Remove the  `devEnvironments.startTimeout` field from the Che Cluster CR: `kubectl edit checluster eclipse-che -n eclipse-che`
```diff
Spec:
(...)
  Dev Environments:
    Default Components:
      Container:
        Image:           quay.io/devfile/universal-developer-image:ubi8-38da5c2
        Source Mapping:  /projects
      Name:              universal-developer-image
    Default Editor:      che-incubator/che-code/insiders
    Default Namespace:
      Auto Provision:                      true
      Template:                            <username>-che
    Disable Container Build Capabilities:  true
    Seconds Of Inactivity Before Idling:   1800
    Seconds Of Run Before Idling:          -1
-    startTimeoutSeconds:                         420s
    Storage:
      Pvc Strategy:  per-user
  Git Services:
  Networking:
    Auth:
      Gateway:
        Config Labels:
          App:        che
          Component:  che-gateway-config
```
6. Verify that the `config.workspace.progressTimeout` from the DWOC is now gone: `kubectl describe dwoc -n eclipse-che`

```YAML
Name:         devworkspace-config
Namespace:    eclipse-che
Labels:       <none>
Annotations:  <none>
API Version:  controller.devfile.io/v1alpha1
Config:
  Workspace:
Kind:  DevWorkspaceOperatorConfig
Metadata:
(...)
```


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
